go-controller/hybrid-overlay/pkg/controller/node.go

```diff
@@ -3,33 +3,121 @@ package controller
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog"
 )
 
-// StartNode creates and starts the hybrid overlay node controller
-func StartNode(nodeName string, kube kube.Interface, wf *factory.WatchFactory) error {
-	klog.Infof("Starting hybrid overlay node...")
-	node, err := NewNode(kube, nodeName)
+// The nodeController interface is implemented by the os-specific code
+type nodeController interface {
+	AddPod(*kapi.Pod) error
+	DeletePod(*kapi.Pod) error
+	AddNode(*kapi.Node) error
+	DeleteNode(*kapi.Node) error
+	RemoveStaleNodes([]*kapi.Node)
+	RemoveStalePods([]*kapi.Pod)
+}
+
+// Node is a node controller and it's informers
+type Node struct {
+	controller   nodeController
+	nodeInformer informer.EventHandler
+	podInformer  informer.EventHandler
+}
+
+// NewNode Returns a new Node
+func NewNode(
+	kube kube.Interface,
+	nodeName string,
+	nodeInformer cache.SharedIndexInformer,
+	podInformer cache.SharedIndexInformer,
+) (*Node, error) {
+	controller, err := newNodeController(kube, nodeName)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	return node.Start(wf)
+	n := &Node{controller: controller}
+	n.nodeInformer = informer.NewDefaultEventHandler("node", nodeInformer,
+		func(obj interface{}) error {
+			node, ok := obj.(*kapi.Node)
+			if !ok {
+				return fmt.Errorf("object is not a node")
+			}
+			return n.controller.AddNode(node)
+		},
+		func(obj interface{}) error {
+			node, ok := obj.(*kapi.Node)
+			if !ok {
+				return fmt.Errorf("object is not a node")
+			}
+			return n.controller.DeleteNode(node)
+		},
+	)
+	n.podInformer = informer.NewDefaultEventHandler("pod", podInformer,
+		func(obj interface{}) error {
+			pod, ok := obj.(*kapi.Pod)
+			if !ok {
+				return fmt.Errorf("object is not a pod")
+			}
+			return n.controller.AddPod(pod)
+		},
+		func(obj interface{}) error {
+			pod, ok := obj.(*kapi.Pod)
+			if !ok {
+				return fmt.Errorf("object is not a pod")
+			}
+			return n.controller.DeletePod(pod)
+		},
+	)
+	return n, nil
 }
 
-// nodeChanged returns true if any relevant node attributes changed
-func nodeChanged(node1 *kapi.Node, node2 *kapi.Node) bool {
-	cidr1, nodeIP1, drMAC1, _ := getNodeDetails(node1)
-	cidr2, nodeIP2, drMAC2, _ := getNodeDetails(node2)
-	return !reflect.DeepEqual(cidr1, cidr2) || !reflect.DeepEqual(nodeIP1, nodeIP2) || !reflect.DeepEqual(drMAC1, drMAC2)
+// Run starts the controller
+func (n *Node) Run(stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	klog.Info("Starting Hybrid Overlay Node Controller")
+
+	klog.Info("Starting workers")
+	go func() {
+		err := n.nodeInformer.Run(informer.DefaultNodeInformerThreadiness, stopCh)
+		if err != nil {
+			klog.Error(err)
+		}
+	}()
+	go func() {
+		err := n.podInformer.Run(informer.DefaultInformerThreadiness, stopCh)
+		if err != nil {
+			klog.Error(err)
+		}
+	}()
+
+	nodeLister := listers.NewNodeLister(n.nodeInformer.GetIndexer())
```

Comment:
minor nit: want to add a `Lister()` function to the `eventHandler` interface?

Reply:
I did think about that, but the lack of generics put me off as I can't return a `NodeLister`.
It would have to be a `Lister` and then caller would need to make a type assertion etc...

Reply:
ugh, good point.